nevergrad/functions/gym/gym_multi.py

```diff
@@ -0,0 +1,198 @@
+        self.discrete = discrete
+
+    def env_names(self):
+        return GYM_ENV_NAMES
+
+    def controllers(self):
+        return CONTROLLERS
+
+    def discretize(self, a):
+        probabilities = np.exp(a - max(a))
+        probabilities = probabilities / sum(probabilities)
+        return int(np.random.multinomial(1, probabilities)[0])
+
+    def neural(self, x: np.ndarray, o: np.ndarray):
+        x = x.reshape(self.policy_shape)
+        if self.control == "linear":
+            output = np.matmul(o.ravel(), x[1:, :])
+            output += x[0]
+            return output.reshape(self.output_shape)
+        first_size = self.num_neurons * (self.input_dim + 1)
+        first_matrix = x[:first_size].reshape(self.input_dim + 1, self.num_neurons)
+        second_matrix = x[first_size:].reshape(self.num_neurons, self.output_dim)
+        return np.matmul(
+            np.tanh(np.matmul(o.ravel(), first_matrix[1:]) + first_matrix[0]), second_matrix
+        ).reshape(self.output_shape)
+
+    def gym_multi_function(self, x: np.ndarray):
+        reward = 0.0
+        num_simulations = 7 if self.control != "conformant" else 1
+        for seed in range(num_simulations):
+            reward += self.gym_simulate(x, seed=seed)
+        return reward / num_simulations
+
+    def gym_simulate(self, x: np.ndarray, seed: int = 0):
+        assert seed == 0 or self.control != "conformant"
+        env = self.env
+        env.seed(seed=seed)
+        o = env.reset()
+        control = self.control
+        if "conformant" in control:
+            return self.gym_conformant(x)
+        if "scrambled" in control:
+            np.random.RandomState(1234).shuffle(x)
+        if "noisy" in control:
+            x = x + 0.01 * np.random.RandomState(1234).normal(size=x.shape)
+        reward = 0.0
+        for i in range(self.num_time_steps):
+            if self.discrete_input:
+                obs = np.zeros(shape=self.input_dim)
+                obs[o] = 1
+                o = obs
+            o = np.asarray(o)
+            a = self.neural(x, o)
+            if self.discrete:
+                a = self.discretize(a)
+            else:
+                a = self.action_type(a)
+            try:
+                assert type(a) == self.action_type
+                o, r, done, _ = env.step(a)  # Outputs = observation, reward, done, info.
+            except AssertionError:  # Illegal action.
+                return 1e20 / (1.0 + i)  # We encourage late failures rather than early failures.
+            reward += r
+            if done:
+                break
+        return reward
+
+    def gym_conformant(self, x: np.ndarray):
+        reward = 0.0
+        for i, a in enumerate(10.0 * x):
+            if type(a) == np.float64:
+                a = np.asarray((a,))
+                if self.discrete:
+                    a = self.discretize(a)
+                else:
+                    try:
+                        if type(a) != self.action_type:
```

Comment:
this is ugly, why do we need this?

Reply:
Because if x is a np.ndarray, then type(x)(x)  does not do what we expect.


import numpy as np
x=np.asarray([1,2,3])
type(x)(x)

array([[[ 0.00000000e+000, -4.33378621e-311,  2.20885204e+161],
        [ 6.01334515e-154,  6.81701550e-313,  4.17201348e-309]]])

(e.g. play with <<python -c "import numpy as np; x=np.asarray([1,2,3]); print(type(x)(x))" >>

Reply:
Hum with the various if then else above we can actually remove this altogether ==> done.